demos/immersed_fem/immersed_fem.py.rst

```diff
@@ -90,8 +126,212 @@ you can type the following command in the terminal
+   version of Gmsh and DMPlex, the Gmsh option ``-format msh2`` may be omitted.
+
+Using the Gmsh API
+------------------
+
+We can alternatively use python commands enabled through the Gmsh API to build, save, and
+read the mesh into Firedrake from within a python script. This allows for parameter flexibility
+and improved readibility of the mesh generation code.
+
+We first need to ``initialize`` the Gmsh API and create a new empty mesh model. 
+
+.. code-block:: none
+
+   gmsh.initialize()
+   model = gmsh.model
+   model.add("gmsh_api_demo")
+
+As before, we define the four rectangle corner points and target element size.
+
+.. code-block:: none
+
+   rectangle_points = [
+   model.geo.addPoint(x0 - L/2, y0 + W/2, z0, dx_rec, 1), # top left
+   model.geo.addPoint(x0 - L/2, y0 - W/2, z0, dx_rec, 2), # bottom left
+   model.geo.addPoint(x0 + L/2, y0 - W/2, z0, dx_rec, 3), # bottom right
+   model.geo.addPoint(x0 + L/2, y0 + W/2, z0, dx_rec, 4)  # top right
+   ]
+
+Then, we define 5 points to describe a circle.
+
+.. code-block:: none
+
+   center = model.geo.addPoint(x0, y0, z0, 5)
+   circle_points = [
+   model.geo.addPoint(x0 - R, y0, z0, dx_disc, 6),
+   model.geo.addPoint(x0, y0 + R, z0, dx_disc, 7),
+   model.geo.addPoint(x0 + R, y0, z0, dx_disc, 8),
+   model.geo.addPoint(x0, y0 - R, z0, dx_disc, 9)
+   ]
+
+Then, we create 8 edges: 4 for the rectangle and 4 for the circle.
+
+.. code-block:: none
+
+   rectangle_lines = [
+   model.geo.addLine(rectangle_points[0], rectangle_points[1], 1), # left
+   model.geo.addLine(rectangle_points[1], rectangle_points[2], 2), # bottom
+   model.geo.addLine(rectangle_points[2], rectangle_points[3], 3), # right
+   model.geo.addLine(rectangle_points[3], rectangle_points[0], 4)  # top
+   ]
+
+   circle_arcs =[
+   model.geo.addCircleArc(circle_points[0], center, circle_points[1], 5),
+   model.geo.addCircleArc(circle_points[1], center, circle_points[2], 6),
+   model.geo.addCircleArc(circle_points[2], center, circle_points[3], 7),
+   model.geo.addCircleArc(circle_points[3], center, circle_points[0], 8)
+   ]
+
+We then combine the edges into a closed loop for both the rectangle and
+the circle used to define a surface for the area outside and inside the 
+inscribed circle, respectively. In the ``addPlaneSurface`` function by 
+convention, the first ``Curve Loop`` defines the outer boundary and
+anything after in the list is treated as a hole in the domain. These need
+to be registered to the model with ``syncronize`` before we can use them.
+
+.. code-block:: none
+
+   rectangle_loop = model.geo.addCurveLoop(rectangle_lines, 9)
+   circle_loop = model.geo.addCurveLoop(circle_arcs, 10)
+
+   punched_surface = model.geo.addPlaneSurface([rectangle_loop,circle_loop], 1)
+   circle_surface = model.geo.addPlaneSurface([circle_loop], 2)
+   model.geo.synchronize()
+   
+Finally, we group together some edges and define ``Physical`` entities.
```

Comment:
The numbers 1 and 2 after the opening brackets below indicate the dimension of the physical groups.